pkg/app/piped/analysisprovider/metrics/prometheus/prometheus.go

```diff
@@ -43,77 +43,122 @@ type Provider struct {
 	logger  *zap.Logger
 }
 
-func NewProvider(address string, timeout time.Duration, logger *zap.Logger) (*Provider, error) {
+func NewProvider(address string, opts ...Option) (*Provider, error) {
+	if address == "" {
+		return nil, fmt.Errorf("address is required")
+	}
 	client, err := api.NewClient(api.Config{
 		Address: address,
 	})
 	if err != nil {
 		return nil, err
 	}
 
-	if timeout == 0 {
-		timeout = defaultTimeout
+	p := &Provider{
+		api:     v1.NewAPI(client),
+		timeout: defaultTimeout,
+		logger:  zap.NewNop(),
+	}
+	for _, opt := range opts {
+		opt(p)
 	}
+	return p, nil
+}
 
-	return &Provider{
-		api:     v1.NewAPI(client),
-		timeout: timeout,
-		logger:  logger.With(zap.String("analysis-provider", ProviderType)),
-	}, nil
+type Option func(*Provider)
+
+func WithTimeout(timeout time.Duration) Option {
+	return func(p *Provider) {
+		p.timeout = timeout
+	}
+}
+
+func WithLogger(logger *zap.Logger) Option {
+	return func(p *Provider) {
+		p.logger = logger.Named("prometheus-provider")
+	}
 }
 
 func (p *Provider) Type() string {
 	return ProviderType
 }
 
-func (p *Provider) RunQuery(ctx context.Context, query string, expected config.AnalysisExpected) (bool, error) {
+func (p *Provider) RunQuery(ctx context.Context, query string, queryRange metrics.QueryRange, evaluator metrics.Evaluator) (bool, error) {
+	if err := queryRange.Validate(); err != nil {
+		return false, err
+	}
+
 	ctx, cancel := context.WithTimeout(ctx, p.timeout)
 	defer cancel()
 
 	p.logger.Info("run query", zap.String("query", query))
 	// TODO: Use HTTP Basic Authentication with the username and password when needed.
-	response, warnings, err := p.api.Query(ctx, query, time.Now())
+	response, warnings, err := p.api.QueryRange(ctx, query, v1.Range{
+		Start: queryRange.From,
+		End:   queryRange.To,
+		Step:  queryRange.Step,
+	})
 	if err != nil {
 		return false, err
 	}
 	for _, w := range warnings {
 		p.logger.Warn("non critical error occurred", zap.String("warning", w))
 	}
-	return p.evaluate(expected, response)
+	return p.evaluate(evaluator, response)
 }
 
-func (p *Provider) evaluate(expected config.AnalysisExpected, response model.Value) (bool, error) {
-	if err := expected.Validate(); err != nil {
+func (p *Provider) evaluate(evaluator metrics.Evaluator, response model.Value) (bool, error) {
```

Comment:
Can we move this function to be an independent one and add some tests for it?

Reply:
@nghialv For sure for sure. Give me a kind of time to do so.

Reply:
@nghialv Made it independent just now. And I'm looking to work on tests for it when I address zero value stuff: https://github.com/pipe-cd/pipe/issues/780 https://github.com/pipe-cd/pipe/issues/1683